File: Java/opentraceability/src/test/java/opentraceability/mappers/MasterDataTests.java
package opentraceability.mappers;

import opentraceability.Setup;
import opentraceability.interfaces.IVocabularyElement;
import opentraceability.models.events.EPCISDocument;
import opentraceability.models.masterdata.Location;
import opentraceability.models.masterdata.TradeItem;
import opentraceability.models.masterdata.TradingParty;
import opentraceability.utility.DataCompare;
import opentraceability.utility.EmbeddedResourceLoader;
import org.junit.jupiter.api.Test;

public class MasterDataTests {

    public MasterDataTests() throws Exception {
        Setup.Initialize();
    }

    @Test
    public void GS1WebVocab_Products() throws Exception {
        String file = "gs1-vocab-products01.json";

        // read object events from test data specified in the file argument
        String json = ReadTestData(file);

        // map into a trade item
        TradeItem tradeitem = (TradeItem)OpenTraceabilityMappers.MasterData.GS1WebVocab.map(TradeItem.class, json);

        // map back into json
        String jsonAfter = OpenTraceabilityMappers.MasterData.GS1WebVocab.map(tradeitem);

        // compare the JSON
        DataCompare.CompareJSON(json, jsonAfter);
    }

    @Test
    public void GS1WebVocab_Location() throws Exception {
        String file = "gs1-vocab-locations01.json";

        // read object events from test data specified in the file argument
        String json = ReadTestData(file);

        // map into a trade item
        Location loc = (Location)OpenTraceabilityMappers.MasterData.GS1WebVocab.map(Location.class, json);

        // map back into json
        String jsonAfter = OpenTraceabilityMappers.MasterData.GS1WebVocab.map(loc);

        // compare the JSON
        DataCompare.CompareJSON(json, jsonAfter);
    }

    @Test
    public void GS1WebVocab_TradingParty() throws Exception {
        String file = "gs1-vocab-tradingparties01.json";

        // read object events from test data specified in the file argument
        String json = ReadTestData(file);

        // map into a trade item
        TradingParty tp = (TradingParty)OpenTraceabilityMappers.MasterData.GS1WebVocab.map(TradingParty.class, json);

        // map back into json
        String jsonAfter = OpenTraceabilityMappers.MasterData.GS1WebVocab.map(tp);

        // compare the JSON
        DataCompare.CompareJSON(json, jsonAfter);
    }

    @Test
    public void GS1WebVocab_EPCISDocument() throws Exception {

        String file = "testserver_advancedfilters.jsonld";

        // read object events from test data specified in the file argument
        String json = ReadTestData(file);

        // map into a trade item
        EPCISDocument doc = OpenTraceabilityMappers.EPCISDocument.JSON.map(json, true);

        // foreach master data item
        for (var md: doc.masterData)
        {
            String jsonMD = OpenTraceabilityMappers.MasterData.GS1WebVocab.map(md);

            IVocabularyElement mdAfter = OpenTraceabilityMappers.MasterData.GS1WebVocab.map(md.getClass(), jsonMD);

            // map back into json
            String jsonMDAfter = OpenTraceabilityMappers.MasterData.GS1WebVocab.map(mdAfter);

            // compare the JSON
            DataCompare.CompareJSON(jsonMD, jsonMDAfter);
        }
    }

    String ReadTestData(String file)
    {
        EmbeddedResourceLoader loader = new EmbeddedResourceLoader();
        String str = loader.readString(Setup.class, "/tests/" + file);
        return str;
    }
}

File: Java/OpenTraceability/src/test/kotlin/opentraceability/event/MasterDataTests.kt
package opentraceability.event

import opentraceability.OpenTraceabilityTests
import opentraceability.gdst.MasterData.GDSTLocation
import opentraceability.interfaces.IVocabularyElement
import opentraceability.mappers.OpenTraceabilityMappers
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource
import kotlin.reflect.KClass


class MasterDataTests {

    @Test
    //@ParameterizedTest(name = "{0}")
    //@ValueSource(strings = ["gs1-vocab-products01.json"])
    fun gs1WebVocabProducts() {
        val json = OpenTraceabilityTests.readTestData("gs1-vocab-products01.json")
        val tradeitem = OpenTraceabilityMappers.MasterData.GS1WebVocab.map(opentraceability.models.masterdata.TradeItem::class, json)
        val jsonAfter = OpenTraceabilityMappers.MasterData.GS1WebVocab.map(tradeitem)
        OpenTraceabilityTests.compareJSON(json, jsonAfter)
    }

    @Test
    //@ParameterizedTest(name = "{0}")
    //@ValueSource(strings = ["gs1-vocab-locations01.json"])
    fun gs1WebVocabLocation() {
        val json = OpenTraceabilityTests.readTestData("gs1-vocab-locations01.json")
        val loc = OpenTraceabilityMappers.MasterData.GS1WebVocab.map(opentraceability.gdst.MasterData.GDSTLocation::class, json)
        val jsonAfter = OpenTraceabilityMappers.MasterData.GS1WebVocab.map(loc)
        OpenTraceabilityTests.compareJSON(json, jsonAfter)
    }

    @Test
    //@ParameterizedTest(name = "{0}")
    //@ValueSource(strings = ["gs1-vocab-tradingparties01.json"])
    fun gs1WebVocabTradingParty() {
        val json = OpenTraceabilityTests.readTestData("gs1-vocab-tradingparties01.json")
        val tp = OpenTraceabilityMappers.MasterData.GS1WebVocab.map(opentraceability.models.masterdata.TradingParty::class, json)
        val jsonAfter = OpenTraceabilityMappers.MasterData.GS1WebVocab.map(tp)
        OpenTraceabilityTests.compareJSON(json, jsonAfter)
    }

    @Test
    //@ParameterizedTest(name = "{0}")
    //@ValueSource(strings = ["testserver_advancedfilters.jsonld"])
    fun gs1WebVocabEPCISDocument() {
        val json = OpenTraceabilityTests.readTestData("testserver_advancedfilters.jsonld")
        val doc = OpenTraceabilityMappers.EPCISDocument.JSON.map(json)
        for (md in doc.masterData) {
            val jsonMD = OpenTraceabilityMappers.MasterData.GS1WebVocab.map(md)
            val mdAfter = OpenTraceabilityMappers.MasterData.GS1WebVocab.map(md::class, jsonMD)
            val jsonMDAfter = OpenTraceabilityMappers.MasterData.GS1WebVocab.map(mdAfter)
            OpenTraceabilityTests.compareJSON(jsonMD, jsonMDAfter)
        }
    }
}